content/blog/2025-12-16-facade-layout-compiler.md

```diff
@@ -0,0 +1,409 @@
+- **Parser:** Uses `sync_to_row()` to skip to the next row declaration after errors
+
+This required careful state management to ensure the compiler doesn't report spurious follow-on errors while still catching genuine independent errors.
+
+### 6. Comprehensive Test Coverage
+
+The most time-consuming aspect was ensuring comprehensive test coverage. My initial tests covered obvious paths but missed edge cases like:
+
+- `E*0` in various positions
+- Mixed case symbols
+- Comments in unusual positions
+- Unicode in comments (valid) vs. Unicode in code (invalid)
+- Multiple errors in the same program
+
+I used an LLM as a test ideation tool: after implementing a feature, I asked it to propose adversarial inputs and edge cases. I then checked and determined the expected behavior and added these cases to my test suite.
+
+---
+
+## Were you successful?
+
+### Quantitative Evaluation
+
+The testing results printout is available in [test.txt](https://github.com/itingtsai/Facade_Layout_Compiler/blob/main/test/test.txt)
+
+Comprehensive testing is hard. There is no existing benchmark. After implementing and validating each new rule, I use an LLM for test ideation (e.g., "Give me edge cases to test user inputs containing characters outside the defined DSL elements.") and then manually determine and verify the expected behavior. This approach helped broaden test coverage to corner cases that I wouldn't have considered. The categorized test suite made it easy to verify that each feature worked in isolation and in combination.
+
+### Test Suite Composition
+
+The test suite comprises 146 facade programs organized into 22 categories:
+
+| Category | Tests | Description |
+|----------|-------|-------------|
+| Basic | 5 | Simple valid inputs |
+| Syntax | 11 | Syntax variations (compact, repeat, case) |
+| Comments | 4 | Comment handling |
+| AutoFill | 5 | Auto-fill behavior |
+| RowNum | 5 | Row number handling |
+| Rules | 4 | Rule declarations |
+| Door | 8 | Door structural validation |
+| Chimney | 8 | Chimney connectivity validation |
+| WinSpace | 4 | Window spacing rule |
+| BadChar | 18 | Invalid character detection |
+| BadSymbol | 6 | Invalid symbol detection |
+| SyntaxErr | 18 | Syntax error handling |
+| Edge | 10 | Edge cases |
+| Complex | 4 | Complex valid facades |
+| Boundary | 4 | Boundary conditions |
+| WS | 5 | Whitespace edge cases |
+| BadSymbol2 | 6 | Additional invalid symbol tests |
+| Stress | 5 | Stress tests (large grids, many rows) |
+| Recovery | 3 | Error recovery |
+| Combo | 4 | Feature combinations |
+| ZeroRepeat | 9 | Zero-count repeat expressions |
+
+All 146 tests pass, demonstrating:
+
+- **Functional correctness:** The compiler accepts valid programs and rejects invalid ones
+- **Rule enforcement precision:** Architectural constraints are enforced exactly as specified
+- **Robustness:** The compiler handles syntactic variations, malformed input, and edge cases gracefully
```

Comment:
Please delete this list (for the same "marketing" reasons as above). You can just say that the test pass.